File: its-my-training-day/leetcode-top-75/questions/strings/sliding-window/minimum-window-substring/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const minimumWindowSubstring = (s: string, t: string): string => {
  if (t.length === 0) return "";
  const target: Record<string, number> = {};
  for (let i = 0; i < t.length; i += 1) target[t[i]] = (target[t[i]] || 0) + 1;
  const need = Object.keys(target).length;
  let formed = 0;
  const window: Record<string, number> = {};
  let left = 0;
  let bestLen = 1 << 30;
  let bestL = 0;
  let bestR = 0;
  for (let right = 0; right < s.length; right += 1) {
    const ch = s[right];
    window[ch] = (window[ch] || 0) + 1;
    if (target[ch] !== undefined && window[ch] === target[ch]) formed += 1;
    while (left <= right && formed === need) {
      if (right - left + 1 < bestLen) {
        bestLen = right - left + 1;
        bestL = left;
        bestR = right;
      }
      const leftCh = s[left];
      window[leftCh] -= 1;
      if (target[leftCh] !== undefined && window[leftCh] < target[leftCh]) formed -= 1;
      left += 1;
    }
  }
  if (bestLen === (1 << 30)) return "";
  return s.slice(bestL, bestR + 1);
};

File: its-my-training-day/leetcode-top-75/questions/strings/sliding-window/minimum-window-substring/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def minimum_window_substring(s: str, t: str) -> str:
    if not t:
        return ""
    target: Dict[str, int] = {}
    i = 0
    while i < len(t):
        ch = t[i]
        target[ch] = target.get(ch, 0) + 1
        i += 1
    need = len(target)
    formed = 0
    window: Dict[str, int] = {}
    left = 0
    best_len = 1 << 30
    best = (0, 0)
    right = 0
    while right < len(s):
        ch = s[right]
        window[ch] = window.get(ch, 0) + 1
        if ch in target and window[ch] == target[ch]:
            formed += 1
        while left <= right and formed == need:
            if right - left + 1 < best_len:
                best_len = right - left + 1
                best = (left, right)
            left_ch = s[left]
            window[left_ch] -= 1
            if left_ch in target and window[left_ch] < target[left_ch]:
                formed -= 1
            left += 1
        right += 1
    if best_len == 1 << 30:
        return ""
    return s[best[0]:best[1] + 1]
